lib/Doctrine/ORM/UnitOfWork.php

```diff
@@ -2339,28 +2345,50 @@ public function createEntity($className, array $data, &$hints = [])
+                    //       A#id = B(to-one)
+                    //       B#id = C(to-one)
+                    //       C#id = D(to-one)
+                    //       E#id = E(to-one)
+                    //       E#id = composite scalar
+                    //       that is a mess, but it is the real world scenario
+                    // @TODO we need to therefore have a generic utility that, given a scalar identifier, gives us
+                    //       a "normalized" identifier. This is pretty much the opposite of what the
+                    //       `IdentifierFlattener` does
+                    // @TODO note that we also need to sort the identifier fields here
+
+                    // @TODO Start of the block to be refactored into "make a deep identifier from a flat one"
+
+                    $normalizedAssociatedId = $this->convertFlatIdentifierIntoRealIdentifierFieldValues(
+                        $targetClass,
+                        $associatedId
+                    );
+
                     switch (true) {
                         // We are negating the condition here. Other cases will assume it is valid!
                         case ($hints['fetchMode'][$class->getClassName()][$field] !== FetchMode::EAGER):
-                            $newValue = $this->em->getProxyFactory()->getProxy($targetEntity, $associatedId);
+                            $newValue = $this->em->getProxyFactory()->getProxy($targetEntity, $normalizedAssociatedId);
                             break;
 
                         // Deferred eager load only works for single identifier classes
                         case (isset($hints[self::HINT_DEFEREAGERLOAD]) && !$targetClass->isIdentifierComposite()):
                             // TODO: Is there a faster approach?
-                            $this->eagerLoadingEntities[$targetClass->getRootClassName()][$relatedIdHash] = current($associatedId);
+                            $this->eagerLoadingEntities[$targetClass->getRootClassName()][$relatedIdHash] = current($normalizedAssociatedId);
 
-                            $newValue = $this->em->getProxyFactory()->getProxy($targetEntity, $associatedId);
+                            $newValue = $this->em->getProxyFactory()->getProxy($targetEntity, $normalizedAssociatedId);
                             break;
 
                         default:
                             // TODO: This is very imperformant, ignore it?
-                            $newValue    = $this->em->find($targetEntity, $associatedId);
+                            $newValue    = $this->em->find($targetEntity, $normalizedAssociatedId);
                             // Needed to re-assign original entity data for freshly loaded entity
                             $managedData = $this->originalEntityData[spl_object_hash($newValue)];
                             break;
                     }
 
+                    // @TODO using `$associatedId` here seems to be risky.
```

Comment:
Very risky indeed: this should be a `flattenId(normalizeId(flatId))` instead, in order to guarantee consistency

Reply:
This code cannot be fixed in this patch: possibly related to #6638, #6648, #6649